[RFC] Run bsim tests with twister

ardo-nordic opened the pull request:
Bsim harness in twister has been modified to be able to run tests. The main changes it entailed are as following:

- For bsim tests, compile/build stage needs to be done before tests are run, that requires rearranging tasks in twister pipeline
- Several bsim harness parameters added (optional), as well as option to specify common part of the harness config
- To avoid unnecessary recompiling of tests, in some test suites there are build-only tests that prepare required binaries, and test only ('no_build' option) tests which rely on other build tests to prepare binaries
- All shell scripts that were used to run the tests are no longer needed

Atm, only host is extensively using bsim harness, and thus is the only one affected.

ardo-nordic committed:
scripts: twister: adding running capabilities to bsim harness

Bsim harness now can also be used to run tests.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
scripts: twister: adjusting twister to run Bsim more efficiently

BSIM tests might be reusing the same binary files, so it makes
sense to be able to split tests into build-only, which will
only build and copy executable files, and run only, which will
run them whenever needed. This requires more sequential pipeline
to avoid race conitions.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
scripts: twister: adding overlaying option in harness config

Adding functionality to split harness_config into common and test-case
specific part, which are joined together for any given test.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
scripts: twister: adding bsim Phy options to harness config

Adding Bsim phy-specific options to harness config. Clean up leftover log
files. Adding option to expand extra args with individual values for
different simulated devices. Adding random seed value to every sim device.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: minor update of LL bsim config file(s)

Changing bsim_exe_name values to a list due to respective schema changes.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: host: making tests runnable by twister

Adding necessary changes to testcase.yaml files in all Bsim host tests
to make them directly runnable by twister's bsim harness.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: host: removing shell scripts that run tests

Shell scripts are no longer needed since twister can be used to build
and run host bsim tests.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: host: minor updates related to twister

Increasing test timeouts for better margins. Removing missed obsolete
shell running files.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

aescolar commented:
I'm quite surprised by this PR. Not just because it appeared quite out of the blue for me, but also because there has been quite related work by others (CC @gchwier ) in these area this overlaps with:

* https://github.com/zephyrproject-rtos/zephyr/issues/62649
* https://github.com/zephyrproject-rtos/zephyr/pull/63691
* https://github.com/zephyrproject-rtos/zephyr/pull/66126
* https://github.com/zephyrproject-rtos/zephyr/pull/66187
* https://github.com/zephyrproject-rtos/zephyr/pull/75291

We have had this want for very long, but so far we have not succeeded. So instead of doing the same we have done in previous PRs where we discussed some details and ended up spending time polishing something which was rejected on principal issues, I would suggest we have a talk about it between the more involved and relevant people, try to reach an agreement on the principal, and if we do, then look at the particular details.

ardo-nordic committed:
tests: bsim: host: restoring test timeouts

Chaning timeouts in testcase.yaml files to be the same as they were
in the shell running scripts.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
scripts: twister: fixing compliance checks and minor issues

Fixing minor compliance errors and issues popping up after rebase.
Fixing unit tests.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

aescolar commented:
I just added the 4.2 milestone, as freeze for 4.1 is today, and this is neither ready not urgent. So to make it clear this is not trying to get into this release.

Thalley commented:
@ardo-nordic next time consider opening such PRs as draft until they pass CI to avoid unnecessary noise for reviewers :)

ardo-nordic committed:
tests: bsim: bluetooth: run host tests with twister

With added support of running bsim tests with Twister, separate
building command is no longer need as tests can be build and
run with single command.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

aescolar commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -978,23 +1043,125 @@ def build(self):
+
+    def _generate_commands(self):
+        def rs():
+            return  f'-rs={random.randint(0, 2**10 - 1)}'
```
The harness should not set the random seed on its own, and less to an actual random value.

> We discussed it with @alwa-nordic and it may be better to use the device ID (same as -d) as the random see

In the executable, the random seed already default to a value that depends on the device number. There is no need to do that again.

theob-pro replied:
> In the executable, the random seed already defaults to a value that depends on the device number. There is no need to do that again.

Okay, yeah I was pretty sure there was already something like that implemented but I wanted to test it today to confirm it ^^

Thalley commented on tests/bsim/bluetooth/host/adv/encrypted/css_sample_data/testcase.yaml:
```diff
@@ -1,10 +1,32 @@
+    bsim_test_ids:
+      - central
+      - peripheral
+    bsim_options:
+      - -RealEncryption=1
```
Do we need the `- -`? 

Could `RealEncryption` have been it's own entry in yaml like `real_encryption: true`?

aescolar replied:
> real_encryption: true

I would really not go that way. Which options you can pass to an executable is an unbound and unknown set. They depend on the test, configuration and target, and are not the problem of the twister harness.

ardo-nordic replied:
Yes, it's a pretty long list of arguments which will easily double the size of the schema if we add all of them.

Thalley commented on tests/bsim/bluetooth/host/adv/encrypted/css_sample_data/testcase.yaml:
```diff
@@ -1,10 +1,32 @@
+      - peripheral
+    bsim_options:
+      - -RealEncryption=1
+      - -argstest
+
```
Doens't `-argstest` describe that the next arguments are test arguments? Where are the arguments here?

alwa-nordic replied:
Looks like the common options are prepended to `bsim_options` below. I also think this is awkward.

ardo-nordic replied:
Common section contains common options, allows to shorten the list when there is a lot of duplication so that it's possible to clearly see only the options which are different. However, that's not mandatory to use.

Thalley commented on tests/bsim/bluetooth/host/adv/extended/testcase.yaml:
```diff
@@ -1,20 +1,56 @@
       CONF_FILE=prj_advertiser.conf
-  bluetooth.host.adv.extended.scanner:
+  bluetooth.host.adv.extended.build_scanner:
+    build_only: true
     harness_config:
-      bsim_exe_name: tests_bsim_bluetooth_host_adv_extended_prj_scanner_conf
+      bsim_exe_name:
+        - tests_bsim_bluetooth_host_adv_extended_prj_scanner_conf
     extra_args:
       CONF_FILE=prj_scanner.conf
+  bluetooth.host.adv.extended.ext_adv:
+    no_build: true
+    harness_config:
+      bsim_exe_name:
+        - tests_bsim_bluetooth_host_adv_extended_prj_advertiser_conf
+        - tests_bsim_bluetooth_host_adv_extended_prj_scanner_conf
+      bsim_test_ids:
+        - ext_adv_advertiser
+        - ext_adv_scanner
```
I think it's interesting that you have a test here that is a build-only and one that is a no-build. 

What's preventing this from just being a test that is being built and run?

ardo-nordic replied:
Each test here requires two different executables, one for scanner and one for advertising. Moving them to a separate build only test allows to build them only once and use the executables in subsequent tests without rebuilding.

alwa-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -947,12 +951,70 @@ class Ztest(Test):
+
+    def wait_bsim_ready(self):
+        start_time = time.time()
+        while time.time() - start_time < Bsim.BSIM_READY_TIMEOUT_S:
```
What does this code do?

ardo-nordic replied:
It's here to ensure that executable file needed for the test is compiled and ready.

ardo-nordic replied:
If you for example run a test suite with one build test and a few run-only tests, and have 8 threads for this, run-only tests need this wait mechanism because they'll be attempted to launch at the same time as build test. With one thread, there will be no waiting in this function because when it's going to be called all binaries will be available.

ardo-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -964,12 +969,72 @@ class Ztest(Test):
+        self._start_time = time.time()
+
+    def _get_exe_path(self, index):
+        return self._exe_paths[index if index < len(self._exe_paths) else 0]
```
Serves the purpose of handling the case (quite common one) when the same executable runs on all simulated devices.

ardo-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -964,12 +969,72 @@ class Ztest(Test):
+            [os.path.join(self._bsim_out_path, exe_name) for exe_name in exe_names]
+
+    def clean_exes(self):
+        self._set_start_time()
```
That's essentially the first actual harness function that is executed, at least for tests that are compiling.

ardo-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -978,23 +1043,125 @@ def build(self):
+        files = glob.glob(os.path.join(self._bsim_out_path, '*.log'))
+        # files += glob.glob(os.path.join(self._bsim_out_path, '*.bin'))
+        try:
+            for file in [f for f in files if os.path.getctime(f) > self._start_time]:
+                os.remove(file)
```
Removing the files that were created by the test. Using generators with conditions is a standard practice in Python.

theob-pro replied:
While you are right list comprehension are common practice, mixing the operations (filtering and the iteration) isn't a good practice.

You could do something like that if you want to keep the list comprehension:

```suggestion
            recent_files = [f for f in files if os.path.getctime(f) > self._start_time]
            for file in recent_files:
                os.remove(file)
```

ardo-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -978,23 +1043,125 @@ def build(self):
+
+            for t in threads:
+                t.join(timeout=timeout)
+        except Exception as e:
```
Understandable, but the purpose here is not to crash the whole session regardless of what happens, and use the status to just fail the test, and allow the session of move over to the next one.

ardo-nordic commented on scripts/schemas/twister/testsuite-schema.yaml:
```diff
@@ -180,6 +180,22 @@ schema;scenario-schema:
         "bsim_exe_name":
           type: str
           required: false
+        "bsim_test_ids":
+          type: seq
+          required: false
+          sequence:
+            - type: str
+        "bsim_verbosity":
+          type: int
+          required: false
+        "bsim_sim_length":
+          type: float
+          required: false
+        "bsim_options":
+          type: seq
+          required: false
+          sequence:
+            - type: str
```
Not very fond of it as well, but that's the current design, and I don't want to bloat the scope further.

ardo-nordic commented on tests/bsim/bluetooth/host/id/settings/testcase.yaml:
```diff
@@ -1,10 +1,30 @@
-      - nrf52_bsim/native
-    harness: bsim
+  bluetooth.host.id.settings_dut2:
+    no_build: true
     harness_config:
-      bsim_exe_name: tests_bsim_bluetooth_host_id_settings_prj_conf
+      bsim_test_ids:
+        - dut2
+      bsim_options:
+        - -flash_rm
+  bluetooth.host.id.settings_dut1:
+    no_build: true
+    harness_config:
+      bsim_test_ids:
+        - dut1
```
That's an interesting one. That really goes against the grain of testing when you make a test that relies on the result of a previous test. Ideally, it shouldn't be. So best solution is to redesign this test in my opinion, and not do anything like that again. As it stands now, the test will pass if it's run as a bunch (pipeline queue is lifo, hence reverse order here in testcase.yaml). What I really don't want to do, is to add this very special case to the framework.

PavelVPV replied:
I can say from the mesh perspective. The problem is that there is currently (unless I'm mistaken) no way to restart a bsim executable inside the test (basically the executable itself). For the test cases where we test that mesh correctly restores data from a persistent storage, we need to simulate a device reboot. And this is achieved by running several tests in a row using same simulation id: https://github.com/zephyrproject-rtos/zephyr/blob/91cf42c657afcc613f215f8af2f8e78fc8d0ffa2/tests/bsim/bluetooth/mesh/tests_scripts/persistence/provisioning.sh#L11-L23

aescolar replied:
> That really goes against the grain of testing when you make a test that relies on the result of a previous test. Ideally, it shouldn't be. So best solution is to redesign this test in my opinion, and not do anything like that again

I think there is a bit of a misunderstanding here. These tests where something is run generating flash content, and then something else is run which uses that flash content, are not different tests. They are just sub-steps of the same test.
Meaning, many tests consists of just running a set of devices where all connect to one phy. But that is just the typical way, not the only way.
The tests @alwa-nordic refers to, run 2 or 3 sets of devices with phys in sequence (the whole sequence is one test). But we also have tests in tree running with the EDTT; and tests where only some of the devices executables connect to the Phy.
And that is in tree. Out of tree, we have tests that run postprocessing scripts on the simulation results and other programs to check what has happened.

One option would be to say that, by now, this twister harness does not support running those types of tests, but overall we need to keep supporting them.

alwa-nordic commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -978,23 +1043,125 @@ def build(self):
+
+        def expand_args(dev_id):
+            try:
+                args = [str(eval(v)[dev_id]) if v.startswith('[') else v for v in extra_args]
```
I might have misunderstood what this does when I gave my suggestion. I assumed it's some kind of variable interpolation. Now I see it's for defining separate `bsim_options` per device, right?

I suggest we transpose `bsim_exe_name`, `bsim_test_ids`, and `bsim_options` in the config, so it looks like a script. What do you think? Example below:

```yaml
devices:
    - exe: foo.exe
      bsim_testid: central
      custom_args:
        - -flash=bar.nvm.bin
        - -argstest
        - 10
```

ardo-nordic committed:
scripts: twister: changed bsim test options layout

Changed bsim test options in a way that allows explicit configuration
of simulated devices, including names (test ids), options and exe
names, while still retaining common options and exe name. Removed
random seed generation for devices. Minor review updates.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: host: updates reflecting new schema

Bsim options schema was updated with explicit individual device
configuration, thus updating all test config files in host.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
doc: twister: update bsim harness docs

Updating bsim harness documentation after adding functionality to
run BabbleSim tests.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: converting new test runnable

New bsim test that was added upstream needed to be converted
to be run with Bsim harness.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

theob-pro commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -964,12 +968,77 @@ class Ztest(Test):
+                exe_names[exe_name] = replacer(f'bs_{self.instance.platform.name}_{exe_name}')
+
+        if not exe_names:
+            exe_names[None] = f'bs_{replacer(self.instance.name)}'
```
What's the goal of doing that? Why `None`?
